type.md

```diff
@@ -0,0 +1,175 @@
+(INTEGER 0 4611686018427387903)
+~~~
+
+Function [`TYPE-OF`][type-of] returns the type of given variable. The returned
+result is a [type specifier][type-specifier]. The first element is the type and
+the remaining part is extra information of that type. In this case it is lower
+bound and upper bound. You can safely ignore it for now. Also remember that
+numbers in Lisp have no limit!
+
+Now let's try to [`SETF`][setf] the variable:
+
+~~~lisp
+* (setf *var* "hello")
+"hello"
+
+* (type-of *var*)
+(SIMPLE-ARRAY CHARACTER (5))
+~~~
+
+You see, the type of the same variable is changed to
+[`SIMPLE-ARRAY`][simple-array], with contents of type [`CHARACTER`][character]
+and length 5.
+
+## Type Hierarchy
+
+Inheritance relationship of Lisp types consists a type tree. For example:
+
+~~~lisp
+* (describe 'number)
+COMMON-LISP:NUMBER
+  [symbol]
+
+NUMBER names the built-in-class #<BUILT-IN-CLASS COMMON-LISP:NUMBER>:
+  Class precedence-list: NUMBER, T
+  Direct superclasses: T
+  Direct subclasses: COMPLEX, REAL
+  No direct slots.
+~~~
+
+Function [`DESCRIBE`][describe] shows that type [`NUMBER`][number] inherits type
+`T` and is directly inherited by types [`COMPLEX`][complex] and [`REAL`][real].
+
+Similarly, the precedence list of type [`INTEGER`][integer] is
+[`INTEGER`][integer] <- [`RATIONAL`][rational] <- [`REAL`][real] <-
+[`NUMBER`][number] <- `T`. Type `T` is the root of **all** types.
+
+It might be confusing that [`NUMBER`][number] is a built-in-class. This is
+because Lisp types are implemented as CLOS classes. For example:
+
+~~~lisp
+* (type-of 1234)
+(INTEGER 0 4611686018427387903)
+
+* (class-of 1234)
+#<BUILT-IN-CLASS COMMON-LISP:FIXNUM>
+~~~
+
+Function [`CLASS-OF`][class-of] gives more specific result.
+
+## Work with Types
+
+Function [`TYPEP`][typep] can be used to check if given argument is of given
+type.
+
+~~~lisp
+* (typep 1234 'integer)
+T
+~~~
+
+Function [`SUBTYPEP`][subtypep] can be used to inspect if a type is sub-type of
+another. It returns 2 values:
+- `T, T` means first argument is sub-type of the second one.
+- `NIL, T` means first argument is *not* sub-type of the second one.
+- `NIL, NIL` means "not determined".
+
+For example:
+
+~~~lisp
+* (subtypep 'integer 'number)
+T
+T
+
+* (subtypep 'string 'number)
+NIL
+T
+~~~
+
+Sometimes you may want to perform different actions according to the type of a
+parameter. Macro [`TYPECASE`][typecase] is your friend in this case:
+
+~~~lisp
+* (defun plus1 (arg)
+    (typecase arg
+      (integer (+ arg 1))
+      (string (concatenate 'string arg "1"))
+      (t 'error)))
+PLUS1
+
+* (plus1 100)
+101 (7 bits, #x65, #o145, #b1100101)
+
+* (plus1 "hello")
+"hello1"
+
+* (plus1 'hello)
+ERROR
+~~~
+
+You can also use [`DEFTYPE`][deftype] to define a new type-specifier. The body
+should be a macro checking given argument is of this type (see
+[`DEFMACRO`][defmacro]). For example:
+
+~~~lisp
+* (deftype square (&optional type size)
+  `(and (array ,type (,size ,size))))
```

Comment:
As a reader I'd appreciate more explanations and other examples :]

There is a `and`, but how many type checks ? (1 or 3 ? If 3, I'd write them one per line)

Reply:
I will re-design the `deftype` example.